File: leetcode/dp/265_Paint_House_II/codes/dp.c
int minCostII(int** costs, int costsSize, int* costsColSize) {
    if (costsSize == 0 || *costsColSize == 0) return 0;

    int n = costsSize;
    int k = *costsColSize;

    int* prev = (int*)calloc(k, sizeof(int));

    for (int i = 0; i < n; i++) {
        int* curr = (int*)malloc(k * sizeof(int));
        int min1 = INT_MAX, min2 = INT_MAX, ind1 = -1;

        for (int j = 0; j < k; j++) {
            if (prev[j] < min1) {
                min2 = min1;
                min1 = prev[j];
                ind1 = j;
            } else if (prev[j] < min2) {
                min2 = prev[j];
            }
        }

        for (int j = 0; j < k; j++) {
            curr[j] = costs[i][j] + (j == ind1 ? min2 : min1);
        }

        free(prev);
        prev = curr;
    }

    int result = INT_MAX;
    for (int i = 0; i < k; i++) {
        if (prev[i] < result) result = prev[i];
    }

    free(prev);
    return result;
}

File: leetcode/dp/265_Paint_House_II/codes/dp.py
from typing import List

class Solution:
    def minCostII(self, costs: List[List[int]]) -> int:
        if not costs: 
            return 0

        n, k = len(costs), len(costs[0])
        prev_dp = [0 for _ in range(k)]

        for i in range(n):
            curr_dp = [0 for _ in range(k)]

            min1 = min2 = float("inf")
            ind1 = -1

            for j in range(k):
                if prev_dp[j] < min1:
                    min2 = min1
                    min1 = prev_dp[j]
                    ind1 = j
                elif prev_dp[j] < min2:
                    min2 = prev_dp[j]

            for j in range(k):
                if j == ind1:
                    curr_dp[j] = costs[i][j] + min2
                else:
                    curr_dp[j] = costs[i][j] + min1

            prev_dp = curr_dp

        return min(prev_dp)
